---
title: "Brian_zebosi_final_R_assignment"
author: "zebosi brian"
date: "October 13, 2017"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

# Part 1 
#prackage installed include dplyr, ggplot2, plyr, tidyr, reshape2, gsub, tibble
```{r}
library(dplyr)
library(ggplot2)
library(reshape2)
library(plyr)
library(tidyr)
library(gsub)
library(tibble)
```
#loading the dataset

```{r}
fang <- read.delim("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2017/master/UNIX_Assignment/fang_et_al_genotypes.txt", header = T)
Position<- read.delim("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2017/master/UNIX_Assignment/snp_position.txt", header = T)
```


# Inspection 
# data inspection using commands str, dim, filesize

#fang_et_al_genotypes.txt
```{r}
str(fang)
dim(fang) # combines the number of rows and colunms 
class(fang)# tells about  class type 
#colnames(fang)
print(object.size(fang), units="Mb")#file size 

```

#snp_position.txt
```{r}
############################### Kevin notes: you need to keep consistent with your capital letters (Position)
str(Position)
dim(Position) # combines the number of rows and colunms 
class(Position) # tells about colunm names, class type 
colnames(Position)
print(object.size(Position), units="Mb")
```

# Data processing 
# grouping by maize and teosinite 

# maize grouping 
```{r}
maize_groups <- c("ZMMIL", "ZMMLR", "ZMMMR")
maize <- fang[fang$Group %in% maize_groups, ] # matching the maize groups and then subsetting by square brackets
dim(maize) # check if the match and subsetting is effected
```

#teosinite grouping 
```{r}
teosinte_groups <- c("ZMPBA", "ZMPIL", "ZMPJA") # create vector eith teosinte groups 
teosinte <- fang[fang$Group %in% teosinte_groups, ] # matching the maize groups and then subsetting by square brackets
############################### Kevin notes: this should say teosinte
dim(maize) # check if the match and subsetting is effected
```

#transposing data file

#maize 
```{r}
maize_transpose <-as.data.frame(t(maize[,-c(1:3)])) # transpose and remove the rows 1,2,3 because donot need them 
dim(maize_transpose)# number of colunms and rows


```

# teosinte 
```{r}
teosinte_transpose <- as.data.frame(t(teosinte[,-c(1:3)])) # transpose and remove the rows 1,2,3 because donot need them
dim(teosinte_transpose)

```

#editing the snp file
```{r}
position_final <- subset(Position[,c(1,3,4)])# extract important colunms 1,3,4
dim(position_final)
############################### Kevin notes: Error in subset(position[, c(1, 3, 4)]) : object 'position' not found
```

# sorting the data file 
```{r}
maize_sorted <- maize_transpose[order(row.names(maize_transpose)),]
teosinte_sorted <- teosinte_transpose[order(row.names(teosinte_transpose)),]
position_sorted <- position_final[order(position_final$SNP_ID),]; position_sorted
```

# filtering to remove data with unknown and random chromosome position 
```{r}
chromosome_names <- c(1,2,3,4,5,6,7,8,9,10)
position_sorted_edit <- subset(position_sorted, Chromosome %in% chromosome_names)
dim(position_sorted_edit)
position_sorted_edit[1:3, 1:3]
```

#merging datasets

```{r}
maize_merged <- merge(position_sorted_edit, maize_sorted, by.x ="SNP_ID", by.y ="row.names") #now merge by the SNP_ID column
dim(maize_merged)

teosinte_merged <- merge(position_sorted_edit, teosinte_sorted, by.x ="SNP_ID", by.y ="row.names")  #now merge by the SNP_ID column
dim(teosinte_merged)
```

#Sort by SNP position with ? - ascending order

```{r}
maize_sort_posi_ascend <- arrange(maize_merged, Position) # sort ascendingly by position
maize_sort_posi_ascend[1:9, 1:9] # check the dataset

teosinte_sort_posi_ascend <- arrange(teosinte_merged, Position) #sort ascendingly by position
teosinte_sort_posi_ascend[1:9, 1:9] # check the dataset
``` 

#replacing ? with - library(gtools)
```{r}
maize_merged_dash <-gsub("?", " -", maize_merged)
teosinte_merged_dash <-gsub("?", " -", teosinte_merged)
```


#sort by Sort by SNP position - descending order with -

```{r}
maize_sort_posi_descend_dash <- arrange(maize_merged, desc(Position))
teosinte_sort_posi_descend_dash <- arrange(teosinte_merged, desc(Position))
############################### Kevin notes: these files don't seem to have substituted the dash for the ?
```

#extracting chromosomes 

#chromosome extraction ascending(?)
```{r}
for(i in 1:10){
  chrom_extract_asc<-maize_sort_posi_ascend[maize_sort_posi_ascend$Chromosome ==i,]
  write.csv(chrom_extract_asc, paste("maize_chromosome_ascend.csv",i,".csv",sep=""),row.names=F)}

for(i in 1:10){
  chrom_extract_asc<-teosinte_sort_posi_ascend[teosinte_sort_posi_ascend$Chromosome==i,]
  write.csv(chrom_extract_asc, paste("teosinte_chromosome_ascend.csv",i,".csv",sep=""),row.names=F)}
```

#chromosome extraction descending(-)

```{r}
for(i in 1:10){
  chrom_extract_desc<-maize_sort_posi_descend_dash[maize_sort_posi_descend_dash$Chromosome ==i,]
  write.csv(chrom_extract_desc, paste("maize_chromosome_descend.csv",i,".csv",sep=""),row.names=F)}

for(i in 1:10){
  chrom_extract_desc<-teosinte_sort_posi_descend_dash[teosinte_sort_posi_descend_dash$Chromosome ==i,]
  write.csv(chrom_extract_desc, paste("maize_chromosome_descend",i,".csv",sep=""),row.names=F)}
```
############################### Kevin notes: these files don't seem to have substituted the dash for the ?


#part 2 

# transpose fang file 
```{r}
fang_transpose <-as.data.frame(t(fang[,-c(1:3)]))
dim(fang_transpose)
```

#merge fang and snp_position.txt
```{r}
fang_trans_snp_merged <- merge(Position, fang_transpose, by.x ="SNP_ID", by.y ="row.names") #now merge by the SNP_ID column
```

#visualization - Plot the total number of SNPs in our dataset on each chromosome
```{r}
fang_trans_snp_merged <- merge(Position, fang_transpose, by.x ="SNP_ID", by.y ="row.names") #now merge by the SNP_ID column

fang_trans_snp_merged$Chromosome <- factor(fang_trans_snp_merged$Chromosome, levels = c("1","2","3","4","5","6","7","8","9","10","unknown","multiple"))
ggplot(fang_trans_snp_merged)+geom_bar(aes(fang_trans_snp_merged$Chromosome))+xlab("chromosome")+ylab("total number of SNPs")

############################### Kevin notes: the graph looks good!
```

#groups contribute most of these SNPs
```{r}
ggplot(fang, aes(Group)) + geom_bar()+xlab("Group")+ylab("Total SNPS")
##From the figure we could see that ZMMIL contribute most to the total snps.

############################### Kevin notes: the graph looks good!

```



####Missing data and amount of heterozygosity



#melt the datasets to make them tidy, using a vector with the names of all the SNPs
```{r}
Names<- colnames(fang)[-c(1:3)]
fang_melt <- melt(fang, measure.vars = Names)
colnames(fang_melt)[c(3,4,5)] <- c("Group","SNP_ID", "Allele")
```

# convert missing data or ? to NA
```{r}
fang_melt[ fang_melt == "?/?" ] = NA
head(fang_melt, n=9)
```

# creating colunm for homozyousity 
```{r}
fang_melt$Homozygous <- (fang_melt$Allele =="A/A" | fang_melt$Allele =="C/C" | fang_melt$Allele =="G/G" | fang_melt$Allele =="T/T")
```

#Sort the data by sample ID and group
```{r}
fang_melt_sort <-arrange(fang_melt,Sample_ID,Group)
head(fang_melt_sort, n=9)
############################### Kevin notes: looks good!

```

#plot of the homozgyotes, heterozygotes and missing by sample_id
```{r}

summ_id <- ddply(fang_melt_sort, c("Sample_ID"), summarise, total_homozygotes=sum(Homozygous, na.rm=TRUE), total_heterozygotes=sum(!Homozygous, na.rm=TRUE), missing = sum(is.na(Homozygous)))

summ_id_melt <- melt(summ_id, measure.vars = c("total_homozygotes", "total_heterozygotes", "missing"))

ggplot(summ_id_melt,aes(x = Sample_ID, y = value, fill=variable)) + geom_bar(stat = "identity", position = "stack")
attributes(summ_id_melt)

############################### Kevin notes: the graph looks amazing! Really good coding!

```

#plot of the homozgyotes, heterozygotes and missing by sample_id
```{r}
Group_id <- ddply(fang_melt_sort, c("Group"), summarise, total_homozygotes=sum(Homozygous, na.rm=TRUE), total_heterozygotes=sum(!Homozygous, na.rm=TRUE), missing = sum(is.na(Homozygous)))

Group_id_melt <- melt(Group_id, measure.vars = c("total_homozygotes", "total_heterozygotes", "missing"))

ggplot(Group_id_melt,aes(x = Group, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack")
```

############################### Kevin notes: NIce work Brian, the graphs look really good!





